pkg/gds/admin_test.go

```diff
@@ -0,0 +1,966 @@
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"sort"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	jwt "github.com/golang-jwt/jwt/v4"
+	"github.com/google/uuid"
+	"github.com/trisacrypto/directory/pkg/gds"
+	admin "github.com/trisacrypto/directory/pkg/gds/admin/v2"
+	"github.com/trisacrypto/directory/pkg/gds/config"
+	"github.com/trisacrypto/directory/pkg/gds/models/v1"
+	"github.com/trisacrypto/directory/pkg/gds/tokens"
+	"github.com/trisacrypto/directory/pkg/utils"
+	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+)
+
+func (s *gdsTestSuite) initAdmin(dbPath string) (admin *gds.Admin) {
+	var err error
+	require := s.Require()
+	s.conf.Database = config.DatabaseConfig{
+		URL: "leveldb:///" + dbPath,
+	}
+	admin, s.db, s.tokens, err = gds.NewMockedAdmin(*s.conf)
+	require.NoError(err)
+	return admin
+}
+
+// httpRequest is a helper struct to make it easier to organize all the different
+// parameters required for making an in-code API request.
+type httpRequest struct {
+	method  string
+	path    string
+	headers map[string]string
+	params  map[string]string
+	in      interface{}
+	claims  *tokens.Claims
+}
+
+// makeRequest creates a new HTTP request and returns the gin context along with the
+// http.ResponseRecorder.
+func (s *gdsTestSuite) makeRequest(request *httpRequest) (*gin.Context, *httptest.ResponseRecorder) {
+	var body io.ReadWriter
+	var err error
+	require := s.Require()
+
+	// Encode the JSON request
+	if request.in != nil {
+		body = &bytes.Buffer{}
+		err = json.NewEncoder(body).Encode(request.in)
+		require.NoError(err)
+	} else {
+		body = nil
+	}
+
+	// Construct the HTTP request
+	w := httptest.NewRecorder()
+	c, _ := gin.CreateTestContext(w)
+	if request.claims != nil {
+		c.Set(admin.UserClaims, request.claims)
+	}
+	c.Request = httptest.NewRequest(request.method, request.path, body)
+	c.Request.Header.Add("Content-Type", "application/json")
+	if request.headers != nil {
+		for k, v := range request.headers {
+			c.Request.Header.Add(k, v)
+		}
+	}
+	if request.params != nil {
+		for k, v := range request.params {
+			c.Params = append(c.Params, gin.Param{
+				Key:   k,
+				Value: v,
+			})
+		}
+	}
+	return c, w
+}
+
+// doRequest is a helper function for making an admin API request and retrieving
+// the response.
+func (s *gdsTestSuite) doRequest(fn gin.HandlerFunc, c *gin.Context, w *httptest.ResponseRecorder, reply interface{}) (res *http.Response) {
+	require := s.Require()
+	// Call the admin function and return the HTTP response
+	fn(c)
+	res = w.Result()
+	defer res.Body.Close()
+	if reply != nil {
+		bytes, err := ioutil.ReadAll(res.Body)
+		require.NoError(err)
+		err = json.Unmarshal(bytes, reply)
+		require.NoError(err)
+	}
+	return res
+}
+
+// Test that we get a good response from ProtectAuthenticate.
+func (s *gdsTestSuite) TestProtectAuthenticate() {
+	require := s.Require()
+	a := s.initAdmin(s.dbPath)
+
+	request := &httpRequest{
+		method: http.MethodPost,
+		path:   "/v2/foo",
```

Comment:
Isn't `ProtectAuthenticate` supposed to be a `GET /v2/authenticate`?

Reply:
You're right, thanks for catching it.